server/game/Interfaces.ts

```diff
@@ -19,7 +19,7 @@ export type ITriggeredAbilityProps = ITriggeredAbilityWhenProps | ITriggeredAbil
 export type IReplacementEffectAbilityProps = IReplacementEffectAbilityWhenProps | IReplacementEffectAbilityAggregateWhenProps;
 
 /** Interface definition for addActionAbility */
-export interface IActionAbilityProps<Source = any> extends Omit<IAbilityProps<AbilityContext<Source>>, 'optional'> {
+export type IActionAbilityProps<Source = any> = Exclude<IAbilityPropsWithSystems<AbilityContext<Source>>, 'optional'> & {
```

Comment:
I'm not super familiar with Typescript syntax. What is the purpose of the change from an interface to a type here, and is the '&' symbol meaningful in some way?

Reply:
Yeah there's some nuance, but at the end of the day the behavior is the same.

The goal here is to make it so that the same properties object can't have both `targetResolver` and `immediateEffect` (or one of a couple other properties but we'll narrow the discussion to those two).

Previously, it was legal to do that because the ability properties interface was just a big union:

```typescript
interface IAbilityProps {
    targetResolver: IActionTargetResolver;
    immediateEffect: GameSystem | GameSystem[];
    title: string;
    locationFilter: LocationFilter | LocationFilter[];
    // ... some other stuff ...
}

interface ITriggeredAbilityProps extends IAbilityProps {
    // add some triggered stuff here
}

interface IActionAbilityProps extends IAbilityProps {
    // add some action stuff here
}
```

However, to split things into two groups ("types that use `immediateEffect`" and "types that use `targetResolver`"), I had to break those out into sub-interfaces and then join them back with a type union like this:

```typescript
 interface IAbilityProps {
    title: string;
    locationFilter: LocationFilter | LocationFilter[];
    // ... some other stuff ...
}

interface IImmediateEffectAbilityProps extends IAbilityProps {
    immediateEffect: GameSystem | GameSystem[];
}

interface ITargetResolverAbilityProps<Context> extends IAbilityProps<Context> {
    targetResolver: IActionTargetResolver;
}

// this forces the user to select one version or the other but not both
type IAbilityPropsWithSystems = IImmediateEffectAbilityProps | ITargetResolverAbilityProps;
```

However, since the derived interfaces `ITriggeredAbilityProps` and `IActionAbilityProps` are now extending a _type_ instead of an interface, they themselves have to become types. The & in this context is functionally the same as `extends`, it's used for appending new properties to an existing type:

```typescript
type ITriggeredAbilityProps = IAbilityPropsWithSystems & {
    // add some triggered stuff here
}

type IActionAbilityProps = IAbilityPropsWithSystems & {
    // add some action stuff here
}
```